compiler/optimizations/optimizeForallUnorderedOps.cpp

```diff
@@ -0,0 +1,865 @@
+  se->setSymbol(newFn);
+
+}
+
+static bool hasAcceptableType(Symbol* sym) {
+  Type* t = sym->getValType();
+  return is_bool_type(t) ||
+         is_int_type(t) ||
+         is_uint_type(t) ||
+         is_real_type(t) ||
+         is_imag_type(t) ||
+         is_complex_type(t) ||
+         is_enum_type(t);
+}
+
+static bool isOptimizeableAssignStmt(Expr* stmt, BlockStmt* loop) {
+  Symbol* lhs = NULL;
+  if (CallExpr* call = toCallExpr(stmt))
+    if (call->isPrimitive(PRIM_ASSIGN))
+      if (SymExpr* lhsSe = toSymExpr(call->get(1)))
+        lhs = lhsSe->symbol();
+
+  if (lhs)
+    if (BlockStmt* defInBlock = toBlockStmt(lhs->defPoint->parentExpr))
+      if (isBlockWithinBlock(defInBlock, loop))
+        if (CallExpr* marker = findMarkerNear(stmt))
+          if (hasOptimizationFlag(marker, OPT_INFO_LHS_OUTLIVES_FORALL) &&
+              hasOptimizationFlag(marker, OPT_INFO_FLAG_NO_TASK_PRIVATE))
+            // This last check can be relaxed in the future if the
+            // runtime unordered code supports more cases.
+            // The earlier part already checked that it's a POD type.
+            if (hasAcceptableType(lhs))
+              return true;
+
+  return false;
+}
+
+
+static void transformAssignStmt(Expr* stmt) {
+  CallExpr* call = toCallExpr(stmt);
+
+  INT_ASSERT(call->isPrimitive(PRIM_ASSIGN));
+
+  Symbol* lhs = toSymExpr(call->get(1))->symbol();
+  Symbol* rhs = toSymExpr(call->get(2))->symbol();
+
+  CallExpr* callToRemove = NULL;
+
+  if (rhs->isRef() == false) {
+    // Find a pattern like
+    //
+    // move rhs PRIM_DEREF rhsRef
+    // PRIM_ASSIGN lhs rhs
+    //
+    // in that event, we'll fold the PRIM_DEREF into the getput call
+    // so conceptually do this pre-transformation to
+    //
+    // PRIM_ASSIGN lhs rhsRef
+    //
+    Symbol* rhsRef = NULL;
+    CallExpr* prevCall = toCallExpr(call->prev);
+    if (prevCall != NULL) {
+      if (prevCall->isPrimitive(PRIM_MOVE) ||
+          prevCall->isPrimitive(PRIM_ASSIGN)) {
+        Symbol* prevLhs = toSymExpr(prevCall->get(1))->symbol();
+        if (prevLhs == rhs) {
+          if (CallExpr* rhsCall = toCallExpr(prevCall->get(2))) {
+            if (rhsCall->isPrimitive(PRIM_DEREF))
+              rhsRef = toSymExpr(rhsCall->get(1))->symbol();
+          } else if (SymExpr* rhsSe = toSymExpr(prevCall->get(2))) {
+            if (rhsSe->symbol()->isRef() && prevCall->isPrimitive(PRIM_ASSIGN))
+              rhsRef = rhsSe->symbol();
+          }
+        }
+      }
+    }
+
+    if (rhsRef != NULL && prevCall != NULL) {
+      callToRemove = prevCall;
+      rhs = rhsRef;
+    }
+  }
+
+  if (lhs->isRef() && rhs->isRef()) {
+    SET_LINENO(call);
+    // add the call to getput
+    if (fReportOptimizeForallUnordered)
+      if (call->getModule()->modTag == MOD_USER || developer)
+        USR_PRINT(call, "Optimized assign to be unordered");
+
+    call->insertBefore(new CallExpr(PRIM_UNORDERED_ASSIGN, lhs, rhs));
```

Comment:
Currently the runtime functions only expect 8 byte GETs (we use isNumericType in the module code as a proxy for this.) Do we already check for that somewhere in here? If not could we add a check either here (or I think preferably to codegen near the ugni check so there's a common place to update when the runtime functionality improves)

Reply:
I added it in the function doing the ugni check